File: src/cpp/instruction.cpp
#include "instruction.hpp"

instruction::instruction(){
  type = -1;
  addr_type = 0;
  addr = 0;
}
instruction::instruction(int tipo, int tipodir, int dir){
  type = tipo;
  addr_type = tipodir;
  addr = dir;
}
instruction::~instruction(){
  type = -1;
  addr_type = 0;
  addr = 0;
}

int instruction::get_type() const{
  return type;
}

int instruction::get_addrtype() const{
  return addr_type;
}

int instruction::get_addr() const{
  return addr;
}
    
std::ostream &operator<<(ostream &output, const instruction &instruccion){
  if(instruccion.get_type() == -1)
    output << "INVALID";
  else if(instruccion.get_type() == 12){
    output << "HALT";
  }else{
    switch(instruccion.get_type()) {
      case 0: 
        output << "LOAD ";
        break;
      case 1:
        output << "STORE ";
        break;
      case 2:
        output << "ADD ";
        break;
      case 3:
        output << "SUB ";
        break;
      case 4:
        output << "MUL ";
        break;
      case 5:
        output << "DIV ";
        break;
      case 6:
        output << "READ ";
        break;
      case 7:
        output << "WRITE ";
        break;
      case 8:
        output << "JUMP ";
        break;
      case 9:
        output << "JZERO ";
        break;
      case 10:
        output << "JGTZ ";
        break;
      case 11:
        output << "HALT";
        break;
    }
    switch(instruccion.get_addrtype()){
      case 0:
        output << "=";
        break;
      case 1:
        break;
      case 2:
        output << "*";
        break;
    }
    output << instruccion.get_addr();
  }
  return output;
}

File: src/java/Instruction.java
/**
 * <H1>Clase Aluc</H1>
 * 
 * Esta clase es la encargada del correcto funcionamiento de cada instrucción,
 * así como la asignación de la siguiente ordena a ejecutar. Es teóricamente
 * una mezcla entre una ALU (unidad aritmético lógica) y una UC (unidad de control).
 * 
 * Para más información contacte con el usuario vía e-mail:
 * alu0100881622@ull.edu.es
 * 
 * @author Ernesto Echeverría González
 * @since 20-02-2017
 * @version 1.0.0
 */
public class Instruction {
  private int type;
  private int addrType;
  private int addr;
  
  
  //Prohibir READ 0 y WRITE 0;
  // Creamos la instrucción con dos Strings: el nombre del comando y su operador
  public Instruction(String code, String operator) {
    
    addrType = 0;
    if(operator.startsWith("=")) addrType = 1;
    if(operator.startsWith("*")) addrType = 2;
    
    if(addrType != 0) operator = operator.substring(1);
    
    addr = Integer.parseInt(operator);
    //Asignamos un número por instrucción
    switch(code) {
      case "LOAD":
      case "load":
        type = 0;
        break;
      case "STORE":
      case "store":
        type = 1;
        break;
      case "ADD":
      case "add":
        type = 2;
        break;
      case "SUB":
      case "sub":
        type = 3;
        break;
      case "MUL":
      case "mul":
        type = 4;
        break;
      case "DIV":
      case "div":
        type = 5;
        break;
      case "READ":
      case "read":
        type = 6;
        break;
      case "WRITE":
      case "write":
        type = 7;
        break;
      case "JUMP":
      case "jump":
        type = 8;
        break;
      case "JZERO":
      case "jzero":
        type = 9;
        break;
      case "JGTZ":
      case "jgtz":
        type = 10;
        break;
      case "HALT":
      case "halt":
        type = 11;
        break;
      default:
        type = -2;
    }
    //Limpiamos opciones no válidas
    if((type == 6 || type == 1) && addrType == 1){
      type = -1;
    }
    

  }
  
  //Getters
  public int getType() {
    return type;
  }
  
  public int getAddrType() {
    return addrType;
  }
  
  public int getAddr() {
    return addr;
  }
  
  //Setters
  public void setType(int type) {
    this.type = type;
  }
  
  public void setAddrType(int addrType) {
    this.addrType = addrType;
  }
  
  public void setAddr(int addr) {
    this.addr = addr;
  }
  
  // Método toString
  public String toString() {
    String cad = new String();
    if(type == -1) {
      cad = "Instrucción inválida.";
    }else{
      switch(type) {
        case 0: 
          cad = cad + "LOAD ";
          break;
        case 1:
          cad = cad + "STORE ";
          break;
        case 2:
          cad = cad + "ADD ";
          break;
        case 3:
          cad = cad + "SUB ";
          break;
        case 4:
          cad = cad + "MUL ";
          break;
        case 5:
          cad = cad + "DIV ";
          break;
        case 6:
          cad = cad + "READ ";
          break;
        case 7:
          cad = cad + "WRITE ";
          break;
        case 8:
          cad = cad + "JUMP ";
          break;
        case 9:
          cad = cad + "JZERO ";
          break;
        case 10:
          cad = cad + "JGTZ ";
          break;
        case 11:
          cad = cad + "HALT";
          break;
        default:
      }
      if(type != 11){
        switch(addrType) {
          case 1:
            cad = cad + "=";
            break;
          case 2:
            cad = cad + "*";
            break;
        }
      cad = cad + addr;
      }
    }
    return cad;
  }
  
  
}